Cargo.toml

```diff
@@ -114,3 +114,6 @@ doc = false
 deny-warnings = []
 vendored-openssl = ["openssl/vendored"]
 pretty-env-logger = ["pretty_env_logger"]
+
+[package.metadata.docs.rs]
+rustdoc-args = ["--document-private-items"]
```

Comment:
Should we leave crates.io alone and only do this for `master`s published docs?

Reply:
Sorry I don't get it. Did you mean instead of adding this, we do this in `publish.py` or somewhere else?

Reply:
In the team meeting, we discussed doing a build and publish of cargo's rustdoc so contributors can see the docs for `master` rather than the last one published.

This gives us two copies of the docs with two different target audiences
- docs.rs: for people depending on cargo
  - private items isn't as helpful here
- github pages: for people contributing to cargo
  - private items would be very helpful here

Reply:
Now I recalled... 🙇🏾‍♂️

I can setup that CI script, and mention the link to doc in doc comment.

A challenging question is which URL we prefer?

- https://doc.crates.io/contrib/apidocs
- https://doc.crates.io/apidocs
- https://doc.crates.io/api
- anything else?

Reply:
@ehuss where did we previously publish this?

Reply:
I don't remember, it's possible I'm confusing it with another repo which auto-published.

I don't have a strong opinion on where. Just be careful since https://doc.crates.io/ (and thus https://rust-lang.github.io/cargo/) has several redirects from the old site to https://doc.rust-lang.org/cargo/.

Reply:
A demo is live on https://weihanglo.tw/cargo/contrib/apidoc/cargo/. I chose <https://doc.crates.io/contrib/apidoc/cargo>, as we might publish more crate docs in the future (?). 

No hurry. Take a look whenever you like to.

Reply:
> I chose https://doc.crates.io/contrib/apidoc/cargo, as we might publish more crate docs in the future (?).

I'm imagining the idea of moving some of the testing docs to `cargo-test-support`.  Would you see that being under https://doc.crates.io/contrib/apidoc/cargo or https://doc.crates.io/contrib/apidoc/cargo-test-support?

The advantage I see to putting all of the crates in one spot is that most of them are likely a cohesive set and a contributor will likely want to follow through the whole, independent of the crate boundaries.

Reply:
cargo-test-support maybe? I don't have strong opinion on this. An empty module is also fine. Just want to highlight that, if cargo is a workspace, we will get a list of crates on the left sidebar out-of-the-box 😂

<img width="300" alt="image" src="https://user-images.githubusercontent.com/14314532/186657888-036102a8-93f8-4d3e-a1b3-c85aee3dd7f6.png">

Otherwise, we need to build each crate and hack on [`windows.ALL_CRATES`](https://github.com/rust-lang/rust/blob/34a6cae28e7013ff0e640026a8e46f315426829d/src/librustdoc/html/render/write_shared.rs#L493-L496) in generated crates.js to make it display our crates, though the search-index.js is hard to merge.

Reply:
Yes, this won't be ideal until we have a workspace.  My question was mostly focused on the URL for now as I understood it that the crate was being embedded in the URL .

Reply:
Building docs separately within the same target directory will automatically merge the contents. There shouldn't be a need for a workspace or any hacking.

Reply:
Just checked it. ehuss is always correct 🎉 

I lean to `https://doc.crates.io/contrib/apidoc/<crate>` for each crate's own API doc. We could have an index page on `https://doc.crates.io/contrib/apidoc` if we want by using [`--index-page`](https://doc.rust-lang.org/rustdoc/unstable-features.html#--index-page-provide-a-top-level-landing-page-for-docs) rustdoc flag. 

For usage documentations, all-docs-in-one-crate is a great solution but I don't think it affects URL of API doc.

Reply:
Having each crate's docs outputted to a separate dir/URL means we can't as easily cross-link.  I feel uncertain on how much of a problem that will be.  Today, we would probably just have two spheres of documentation: cargo and cargo-test-support.  I think these can get away with being separate but where I am uncertain is if we split much more out of cargo.

There is a part of me that also wonders about URL stability with crate name changes but I feel like that isn't much of a problem.

Reply:
Thinking it through, I think `apidpc/cargo` is probably the right decision.  Even if we put all cargo-related crates in `apidoc/cargo`, it leaves us open to evolving things in the future vs locking us down by not having space for other names.